ChainSafe/forest
# implement Filecoin.ChainSetHead endpoint

## Summary of changes



Changes introduced in this pull request:

- Add `ChainSetHead` endpoint
- Add `chain set-head` subcommand

### Live testing done:
- Run the local daemon, with an empty keystore passphrase. Let it sync.
```console
$ cargo run --bin forest -- --chain=calibnet --save-token=token
...
2023-05-04T01:26:01.742836Z  INFO forest_chain_sync::tipset_syncer: Successfully synced tipset range: [527905, 527906]
```
- Get a tipset from an earlier epoch using a block explorer. E.g https://calibration.filscan.io/tipset/chain?height=527666 gives us a tipset of `bafy2bzacecaqwjlwim5ylzcpkkktjvz6qz47mrrw5uaohu2nj7c4rnppohpds bafy2bzaced74lwq6emvroxrlcmblf64bo47p35da3immhvfhtfmkv6hd2jrpu`
- Run the new command
```console
$ cargo run --bin forest-cli -- --token=$(cat token) chain set-head bafy2bzacecaqwjlwim5ylzcpkkktjvz6qz47mrrw5uaohu2nj7c4rnppohpds bafy2bzaced74lwq6emvroxrlcmblf64bo47p35da3immhvfhtfmkv6hd2jrpu
```
- Observe that the local head is now behind, and that the daemon validates to catch up. (It may freeze//need restarting after running the command).
```console
$ cargo run --bin forest -- --chain=calibnet --save-token=token
...
2023-05-04T01:31:38.256067Z  INFO forest_chain_sync::chain_muxer: Local node is behind the network, starting BOOTSTRAP from LOCAL_HEAD = 527666 -> NETWORK_HEAD = 527917
...
2023-05-04T01:31:39.530603Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 527666, N blocks = 2
...
2023-05-04T01:32:22.487042Z  INFO forest_chain_sync::tipset_syncer: Validating tipset: EPOCH = 527667, N blocks = 4
```

## Reference issue to close (if applicable)





Closes #2706 

## Other information and links
I've got a low confidence in this change - I observed forest freezing up for O(minute) after running the command, and needed restarting.
That might be acceptable given our usecase is snapshotting, but still.

I couldn't find any way of testing my RPC call other than live testing. Suggestion from the team was checking in some shell to `scripts/calibnet_health_check.sh`, which seems like more of a smoke test than anything.
If it's true that there's no way to hermetically test our node, then I think we should raise a new issue and add it ASAP - this kind of infrastructure is super important both for feature development, confidence, and learning.



## Change checklist




- [x] I have added tests that prove my fix is effective or that my feature works (if possible)
  - We've added a smoke test in `calibnet_health_check.sh`. It's at the bottom, because this is a dev command that could brick the node.
- [x] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.



[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md
